src/Common/src/TypeSystem/Interop/IL/Marshaller.cs

```diff
@@ -772,12 +773,24 @@ protected virtual void EmitMarshalArgumentNativeToManaged()
                 PropagateFromByRefArg(_ilCodeStreams.MarshallingCodeStream, _nativeHome);
             }
 
-            AllocAndTransformNativeToManaged(_ilCodeStreams.MarshallingCodeStream);
+            if (IsNativeByRef && !In)
+            {
+                ReInitManagedTransform(_ilCodeStreams.MarshallingCodeStream);
+            }
+            else
+            {
+                AllocAndTransformNativeToManaged(_ilCodeStreams.MarshallingCodeStream);
+            }
 
             LoadManagedArg(_ilCodeStreams.CallsiteSetupCodeStream);
 
             if (Out)
             {
+                if (IsNativeByRef && !In)
```

Comment:
Even if it is [in], you also may need to allocate. I know you may have potential optimization in mind - if the managed value doesn't get assigned/changed, you can skip the allocation. But this can get tricky and in practice we don't really do it.